File: client-old/src/components/story.ts
import dateformat from "dateformat";
import m from "mithril";

import { humanize } from "../utils";
import { IStory } from "../models/story";

interface IComponentStory {
    story: IStory;
}

export class ComponentStory implements m.ClassComponent<IComponentStory> {
    view(vnode: m.CVnode<IComponentStory>) {
        let authors: m.Children[] = [];

        vnode.attrs.story.authors.forEach((a, i) => {
            authors.push(m(m.route.Link, { "key": `${a.id}-author`, "href": `/author/${a.id}/1` }, a.name));

            if (i != vnode.attrs.story.authors.length - 1) {
                authors.push(", ");
            }
        });

        let origins: m.Children[] = [];

        vnode.attrs.story.origins.forEach((o, i) => {
            origins.push(m(m.route.Link, { "key": `${o.id}-author`, "href": `/origin/${o.id}/1` }, o.name));

            if (i != vnode.attrs.story.origins.length - 1) {
                origins.push(m("span", { "key": `${o.id}-span` }, ", "));
            }
        });

        return m("div", { "panel": true }, [
            m("div", { "head": true, "flex": true }, [
                m("div", { "flex": true, "flex-item": "grow" }, [
                    m("div", { "square": true, "mar-right": "normal" }, [
                        m("div", {
                            "part": "top left",
                            "bg": ((type) => {
                                switch (type) {
                                    case "explicit":
                                        return "black-500";
                                    case "mature":
                                        return "red-500";
                                    case "teen":
                                        return "green-600";
                                    default:
                                        return "gray-700";
                                }
                            })(vnode.attrs.story.square.rating),
                        }),
                        m("div", {
                            "part": "top right",
                            "bg": ((type) => {
                                switch (type) {
                                    case "using":
                                        return "orange-500";
                                    default:
                                        return "gray-700";
                                }
                            })(vnode.attrs.story.square.warnings),
                        }),
                        m("div", {
                            "part": "bottom center",
                            "bg": ((type) => {
                                switch (type) {
                                    case "completed":
                                        return "green-600";
                                    case "in-progress":
                                        return "blue-500";
                                    case "hiatus":
                                        return "purple-500";
                                    default:
                                        return "red-500";
                                }
                            })(vnode.attrs.story.square.state),
                        }),
                    ]),
                    m("div", [
                        m("h3", [
                            m(m.route.Link, { "href": `/story/${vnode.attrs.story.id}/1` }, vnode.attrs.story.name),
                            " by ",
                            authors,
                            // " ",
                            // m(m.route.Link, { "txt": "small", "href": `/story/${vnode.attrs.story.id}/reviews/1` }, "[reviews]"),
                        ]),
                        m("p", origins),
                    ]),
                ]),
                m("p", { "flex-item": "grow", "txt": "right" }, dateformat(vnode.attrs.story.updated, "mmm d, yyyy")),
            ]),
            m("div", { "body": true }, [
                m("p", vnode.attrs.story.summary),
                m("div", { "label-list": true }, [
                    vnode.attrs.story.tags.map((t) => m(m.route.Link, {
                        "key": t.id,
                        "label": true,
                        "href": `/tag/${t.id}/1`,
                        "bg": ((type) => {
                            switch (type) {
                                case "warning":
                                    return "red-500";
                                case "pairing":
                                    return "orange-500";
                                case "character":
                                    return "purple-500";
                                default:
                                    return "gray-700";
                            }
                        })(t.type),
                    }, t.name)),
                ]),
            ]),
            m("div", { "foot": true }, [
                m("div", { "flex": true }, [
                    m("p", { "flex-item": "grow", "txt": "small" }, ``),
                    m("p", { "flex-item": "grow", "txt": "small right" }, `${vnode.attrs.story.language} | ${humanize(vnode.attrs.story.words)} words | ${humanize(vnode.attrs.story.chapters)} chapters`),
                ]),
            ]),
        ]);
    }
}

File: stry-common/src/models/story.rs
use {
    crate::models::{Author, Character, Origin, Pairing, Series, Tag, Warning},
    chrono::{DateTime, Utc},
    std::fmt,
};

#[rustfmt::skip]
#[derive(Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
#[derive(serde::Deserialize, serde::Serialize)]
pub struct Story {
    pub id: String,

    pub name: String,
    pub summary: String,

    pub square: Square,

    pub chapters: u32,
    pub words: u32,

    pub authors: Vec<Author>,
    pub origins: Vec<Origin>,

    pub warnings: Vec<Warning>,
    pub pairings: Vec<Pairing>,
    pub characters: Vec<Character>,
    pub tags: Vec<Tag>,

    pub series: Option<Series>,

    pub created: DateTime<Utc>,
    pub updated: DateTime<Utc>,
}

pub struct StoryRow {
    pub id: String,

    pub name: String,
    pub summary: String,

    pub rating: Rating,
    pub state: State,

    pub created: DateTime<Utc>,
    pub updated: DateTime<Utc>,
}

pub struct StoryPart {
    pub id: String,

    pub name: String,
    pub summary: String,

    pub rating: Rating,
    pub state: State,

    pub chapters: u32,
    pub words: u32,

    pub authors: Vec<Author>,
    pub origins: Vec<Origin>,

    pub warnings: Vec<Warning>,
    pub pairings: Vec<Pairing>,
    pub characters: Vec<Character>,
    pub tags: Vec<Tag>,

    pub created: DateTime<Utc>,
    pub updated: DateTime<Utc>,
}

#[rustfmt::skip]
#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
#[derive(serde::Deserialize, serde::Serialize)]
#[cfg_attr(feature = "types-postgres", derive(postgres_types::ToSql, postgres_types::FromSql))]
#[cfg_attr(feature = "types-postgres", postgres(name = "rating"))]
pub enum Rating {
    #[serde(rename = "explicit")]
    #[cfg_attr(feature = "types-postgres", postgres(name = "explicit"))]
    Explicit,

    #[serde(rename = "mature")]
    #[cfg_attr(feature = "types-postgres", postgres(name = "mature"))]
    Mature,

    #[serde(rename = "teen")]
    #[cfg_attr(feature = "types-postgres", postgres(name = "teen"))]
    Teen,

    #[serde(rename = "general")]
    #[cfg_attr(feature = "types-postgres", postgres(name = "general"))]
    General,
}

impl Rating {
    pub fn title(self) -> &'static str {
        match self {
            Rating::Explicit => "Explicit",
            Rating::Mature => "Mature",
            Rating::Teen => "Teen",
            Rating::General => "General",
        }
    }
}

impl fmt::Display for Rating {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Rating::Explicit => "background--red",
                Rating::Mature => "background--yellow",
                Rating::Teen => "background--green",
                Rating::General => "background--blue",
            }
        )
    }
}

#[cfg(feature = "types-sqlite")]
impl rusqlite::types::FromSql for Rating {
    fn column_result(value: rusqlite::types::ValueRef) -> rusqlite::types::FromSqlResult<Self> {
        value
            .as_str()
            .and_then(|s| match s.to_lowercase().as_str() {
                "explicit" => Ok(Rating::Explicit),
                "mature" => Ok(Rating::Mature),
                "teen" => Ok(Rating::Teen),
                "general" => Ok(Rating::General),
                _ => Err(rusqlite::types::FromSqlError::InvalidType),
            })
    }
}

#[cfg(feature = "types-sqlite")]
impl rusqlite::types::ToSql for Rating {
    fn to_sql(&self) -> rusqlite::Result<rusqlite::types::ToSqlOutput> {
        match self {
            Rating::Explicit => Ok("explicit".into()),
            Rating::Mature => Ok("mature".into()),
            Rating::Teen => Ok("teen".into()),
            Rating::General => Ok("general".into()),
        }
    }
}

#[rustfmt::skip]
#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
#[derive(serde::Deserialize, serde::Serialize)]
#[cfg_attr(feature = "types-postgres", derive(postgres_types::ToSql, postgres_types::FromSql))]
#[cfg_attr(feature = "types-postgres", postgres(name = "state"))]
pub enum State {
    #[serde(rename = "completed")]
    #[cfg_attr(feature = "types-postgres", postgres(name = "completed"))]
    Completed,

    #[serde(rename = "in-progress")]
    #[cfg_attr(feature = "types-postgres", postgres(name = "in-progress"))]
    InProgress,

    #[serde(rename = "hiatus")]
    #[cfg_attr(feature = "types-postgres", postgres(name = "hiatus"))]
    Hiatus,

    #[serde(rename = "abandoned")]
    #[cfg_attr(feature = "types-postgres", postgres(name = "abandoned"))]
    Abandoned,
}

impl State {
    pub fn title(self) -> &'static str {
        match self {
            State::Completed => "Completed",
            State::InProgress => "In Progress",
            State::Hiatus => "Hiatus",
            State::Abandoned => "Abandoned",
        }
    }
}

impl fmt::Display for State {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{}",
            match self {
                State::Completed => "background--green",
                State::InProgress => "background--blue",
                State::Hiatus => "background--purple",
                State::Abandoned => "background--red",
            }
        )
    }
}

#[cfg(feature = "types-sqlite")]
impl rusqlite::types::FromSql for State {
    fn column_result(value: rusqlite::types::ValueRef) -> rusqlite::types::FromSqlResult<Self> {
        value
            .as_str()
            .and_then(|s| match s.to_lowercase().as_str() {
                "completed" => Ok(State::Completed),
                "in-progress" => Ok(State::InProgress),
                "hiatus" => Ok(State::Hiatus),
                "abandoned" => Ok(State::Abandoned),
                invalid => {
                    tracing::debug!("Invalid state value: {}", invalid);
                    Err(rusqlite::types::FromSqlError::InvalidType)
                }
            })
    }
}

#[cfg(feature = "types-sqlite")]
impl rusqlite::types::ToSql for State {
    fn to_sql(&self) -> rusqlite::Result<rusqlite::types::ToSqlOutput> {
        match self {
            State::Completed => Ok("completed".into()),
            State::InProgress => Ok("in-progress".into()),
            State::Hiatus => Ok("hiatus".into()),
            State::Abandoned => Ok("abandoned".into()),
        }
    }
}

#[rustfmt::skip]
#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
#[derive(serde::Deserialize, serde::Serialize)]
pub struct Square {
    pub rating: Rating,
    pub warnings: bool,
    pub state: State,
}
